maple/db.py

```diff
@@ -0,0 +1,195 @@
+"""Support moving data into and out of the database.
+
+In an ideal world, we'd directly store and retrieve objects from the
+`maple.types` module. However, these types don't map perfectly to a relational
+database (e.g. a `Bill` has a list of `Action`s associated with it, and we need
+to represent this with a join.) To that end, we introduce new types with an 'M'
+suffix (for 'Model') that act as a wrapper around the associated object for use
+with the database.
+
+As an example, consider the `ActionM` class. This
+
+"""
+
+
+import logging
+from collections import defaultdict
+from contextlib import contextmanager
+from dataclasses import dataclass
+from pathlib import Path
+from typing import Iterable, Iterator, Type
+
+import peewee as pw
```

Comment:
Do you have an environment (e.g. pip or conda freeze or requirements.txt) spec?

Reply:
Yes, but not in a way that is generally helpful. (All the `nix` files rather rigorously define the environment, but I don't want to make anyone else use Nix.)

I'm happy to make an actual `pyproject.toml` describing the package, and even using `poetry` to then define requirements. Any opposition to `poetry`? (Whatever gets picked, I think we should also just... make a Dockerfile)

Reply:
Sounds good. I'm down for `poetry`.  Let's not merge in the `nix` stuff, in that case.

Reply:
There's now a `pyproject.toml` and associated `poetry.lock`. There's also a `Dockerfile` and associated `docker-compose.yml`.